extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java

```diff
@@ -0,0 +1,571 @@
+        Execs.multiThreaded(taskQueueConfig.getMaxSize(), "k8s-task-runner-%d")
+    );
+    Preconditions.checkArgument(
+        taskQueueConfig.getMaxSize() < Integer.MAX_VALUE,
+        "The task queue bounds how many concurrent k8s tasks you can have, set it to a resonable value"
+    );
+  }
+
+
+  @Override
+  public Optional<InputStream> streamTaskLog(String taskid, long offset)
+  {
+    Optional<InputStream> peonLogs = client.getPeonLogs(new K8sTaskId(taskid));
+    if (!peonLogs.isPresent()) {
+      return Optional.absent();
+    }
+    return Optional.of(peonLogs.get());
+  }
+
+  @Override
+  public ListenableFuture<TaskStatus> run(Task task)
+  {
+    synchronized (tasks) {
+      tasks.computeIfAbsent(
+          task.getId(), k -> new K8sWorkItem(
+              task,
+              exec.submit(() -> {
+                K8sTaskId k8sTaskId = new K8sTaskId(task);
+                try {
+                  JobResponse completedPhase;
+                  Optional<Job> existingJob = client.jobExists(k8sTaskId);
+                  if (!existingJob.isPresent()) {
+                    PeonCommandContext context = new PeonCommandContext(
+                        generateCommand(task),
+                        javaOpts(task),
+                        taskConfig.getTaskDir(task.getId()),
+                        node.isEnableTlsPort()
+                    );
+                    Job job = adapter.fromTask(task, context);
+                    log.info("Job created %s and ready to launch", k8sTaskId);
+                    Pod peonPod = client.launchJobAndWaitForStart(
+                        job,
+                        KubernetesTaskRunnerConfig.toMilliseconds(k8sConfig.k8sjobLaunchTimeout),
+                        TimeUnit.MILLISECONDS
+                    );
+                    log.info("Job %s launched in k8s", k8sTaskId);
+                    completedPhase = monitorJob(peonPod, task, k8sTaskId);
+                  } else {
+                    Job job = existingJob.get();
+                    if (job.getStatus().getActive() == null) {
+                      if (job.getStatus().getSucceeded() != null) {
+                        completedPhase = new JobResponse(job, PeonPhase.SUCCEEDED);
+                      } else {
+                        completedPhase = new JobResponse(job, PeonPhase.FAILED);
+                      }
+                    } else {
+                      // the job is active lets monitor it
+                      completedPhase = monitorJob(task, k8sTaskId);
+                    }
+                  }
+                  TaskStatus status;
+                  if (PeonPhase.SUCCEEDED.equals(completedPhase.getPhase())) {
+                    status = TaskStatus.success(task.getId(), completedPhase.getLocation());
+                  } else {
+                    status = TaskStatus.failure(
+                        task.getId(),
+                        "Task failed %s: " + k8sTaskId
+                    );
```

Comment:
These two blocks handle the completed/failed status we just set above. Can the blocks be merged into the prior blocks? Or, will we come this way more than once if the job ends?

Reply:
This is a weird case.  So the task reports back on completion its status.  But in case the call is missed by the overlord, or in the case when the task process just dies like a `kill -9` dead.  Then we still want to track the job, in the former case the job can still succeed and the overlord callback was missed, in the case of a kill -9 the task would never report back.  So as our last step we grab the status from the job itself.